NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.cpp

```diff
@@ -103,6 +103,68 @@ void LocalLinearLeastSquaresExtrapolator::extrapolate(
                                          *counts);
 }
 
+void LocalLinearLeastSquaresExtrapolator::extrapolate(
+    const unsigned num_components,
+    ExtrapolatableElementCollection const& extrapolatables,
+    GlobalVector const& int_pt_x,
+    std::vector<std::vector<GlobalIndexType>>& chemical_system_index_map)
+{
+    auto const num_nodal_dof_result =
+        _dof_table_single_component.dofSizeWithoutGhosts() * num_components;
+
+    std::vector<GlobalIndexType> ghost_indices;
+    {  // Create num_components times version of ghost_indices arranged by
+       // location. For example for 3 components and ghost_indices {5,6,10} we
+       // compute {15, 16, 17,  18, 19, 20,  30, 31, 32}.
+        auto const& single_component_ghost_indices =
+            _dof_table_single_component.getGhostIndices();
+        auto const single_component_ghost_indices_size =
+            single_component_ghost_indices.size();
+        ghost_indices.reserve(single_component_ghost_indices_size *
+                              num_components);
+        for (unsigned i = 0; i < single_component_ghost_indices_size; ++i)
+        {
+            for (unsigned c = 0; c < num_components; ++c)
+            {
+                ghost_indices.push_back(
+                    single_component_ghost_indices[i] * num_components + c);
+            }
+        }
+    }
+
+    if (!_nodal_values ||
+#ifdef USE_PETSC
+        _nodal_values->getLocalSize() + _nodal_values->getGhostSize()
+#else
+        _nodal_values->size()
+#endif
+            != static_cast<GlobalIndexType>(num_nodal_dof_result))
+    {
+        _nodal_values = MathLib::MatrixVectorTraits<GlobalVector>::newInstance(
+            {num_nodal_dof_result, num_nodal_dof_result, &ghost_indices,
+             nullptr});
+    }
+    _nodal_values->setZero();
+
+    // counts the writes to each nodal value, i.e., the summands in order to
+    // compute the average afterwards
+    auto counts =
+        MathLib::MatrixVectorTraits<GlobalVector>::newInstance(*_nodal_values);
+    counts->setZero();
+
+    auto const size = extrapolatables.size();
+    for (std::size_t i = 0; i < size; ++i)
+    {
+        extrapolateElement(i, num_components, extrapolatables, int_pt_x,
+                           chemical_system_index_map, *counts);
+    }
+
+    MathLib::LinAlg::finalizeAssembly(*_nodal_values);
+
+    MathLib::LinAlg::componentwiseDivide(*_nodal_values, *_nodal_values,
+                                         *counts);
```

Comment:
This part but for one line is equal to the existing implementation. This is a lot of code duplication.